Add "Tips for your first Sentence Transformers project" to the blog (#381)
huggingface/blog: liftoff-tips-for-sentence-transformers.md

---

> title: 'Liftoff! Some tips for your first Sentence Transformers project 🚀'

Reviewer: If the name of the MD file is `liftoff-tips-for-sentence-transformers` then the link for the blog would be `hf.co/blog/liftoff-tips-for-sentence-transformers` which is long and that might be the best for SEO.

Maybe only `tips-for-sentence-transformers` would be better.

thumbnail: /blog/assets/83_st_first_project/thumbnail.png
---

<h1>
    Liftoff! Some tips for your first Sentence Transformers project 🚀
</h1>

<div class="blog-metadata">
    <small>Published June 23, 2022.</small>

>     <a target="_blank" class="btn no-underline text-sm mb-5 font-sans" href="https://github.com/huggingface/blog/blob/main/st-inference-api.md">

Reviewer: ```suggestion
    <a target="_blank" class="btn no-underline text-sm mb-5 font-sans" href="https://github.com/huggingface/blog/blob/main/liftoff-tips-for-sentence-transformers.md">
```

If you change the name of the Markdown file then these would change too.

        Update on GitHub
    </a>
</div>

<div class="author-card">
    <a href="/nimaboscarino"> 
        <img class="avatar avatar-user" src="https://aeiljuispo.cloudimg.io/v7/https://s3.amazonaws.com/moonup/production/uploads/1647889744246-61e6a54836fa261c76dc3760.jpeg?w=200&h=200&f=face" title="Gravatar">
        <div class="bfc">
            <code>nimaboscarino</code>
            <span class="fullname">Nima Boscarino</span>
        </div>
    </a>
</div>

> People who are new to the Machine Learning world often run into two recurring stumbling blocks. The first is choosing the right library to learn, which can be a daunting task when there are so many to pick from. Even once you’ve settled on a library and gone through some tutorials, the next issue is coming up with your first big project and scoping it properly so that you maximize your learning. If you’ve run into those problems, Sentence Transformers is a great library to choose for a number of reasons. In this post I’ll take you through some tips for going from 0 to 100 with a new tool like Sentence Transformers. We’ll also talk about how I built my first ST-powered project, and what I learned along the way. You may also see why you should add Sentence Transformers to your toolkit for your next project 🛠 

Reviewer: ```suggestion
People who are new to the Machine Learning world often run into two recurring stumbling blocks. The first is choosing the right library to learn, which can be daunting when there are so many to pick from. Even once you’ve settled on a library and gone through some tutorials, the next issue is coming up with your first big project and scoping it properly to maximize your learning. If you’ve run into those problems, Sentence Transformers is a great library to choose for a number of reasons. In this post, I’ll take you through some tips for going from 0 to 100 with a new tool like Sentence Transformers. We’ll also talk about how I built my first ST-powered project and what I learned along the way. You may also see why you should add Sentence Transformers to your toolkit for your next project 🛠 
```

## What is Sentence Transformers?

Sentence embeddings? Semantic search? Cosine similarity?!?! 😱 Just a few short weeks ago, these terms were so confusing to me that they made my head spin. I’d heard that [Sentence Transformers](https://www.sbert.net) (ST) was a powerful and versatile library for working with language and image data and I was eager to play around with it, but I was worried that I would be out of my depth.  As it turns out, I couldn’t have been more wrong!

Sentence Transformers is [among the libraries that Hugging Face integrates with](https://huggingface.co/docs/hub/models-libraries), where it’s described with the following:

> Compute dense vector representations for sentences, paragraphs, and images

> In a nutshell, Sentence Transformers answers one question: What if we could treat sentences as points in a multi-dimensional space? This means that ST lets you give it an arbitrary string of text (e.g. “I’m so glad I learned to code with Python!”) and it’ll transform it to a vector, such as `[0.2, 0.5, 1.3, 0.9]`. Another sentence, such as “Python is a great programming language.”, would be transformed to a different vector. These vectors are called “embeddings”, and [they play an important role in Machine Learning](https://medium.com/@b.terryjack/nlp-everything-about-word-embeddings-9ea21f51ccfe).

Reviewer: Could be relevant to mention that both embeddings coexist in the same vector space.

```suggestion
In a nutshell, Sentence Transformers answers one question: What if we could treat sentences as points in a multi-dimensional vector space? This means that ST lets you give it an arbitrary string of text (e.g., “I’m so glad I learned to code with Python!”), and it’ll transform it into a vector, such as `[0.2, 0.5, 1.3, 0.9]`. Another sentence, such as “Python is a great programming language.”, would be transformed into a different vector. These vectors are called “embeddings,” and [they play an essential role in Machine Learning](https://medium.com/@b.terryjack/nlp-everything-about-word-embeddings-9ea21f51ccfe). If these two sentences were embedded with the same model, then both would coexist in the same vector space, allowing for many interesting possibilities.
```

> What makes ST particularly useful is that, once you’ve generated some embeddings, you can use the built-in utility functions to compare how similar one sentence is to another, ***including synonyms!*** 🤯 One way to do this is with [“Cosine Similarity”](https://www.machinelearningplus.com/nlp/cosine-similarity/), and with ST you can skip all the pesky math and just call the *very* handy `util.cos_sim` function to get a score from -1 to 1 that signifies how “similar” the sentences are – the bigger the score is, the more similar the sentences are! 

Reviewer: ```suggestion
What makes ST particularly useful is that, once you’ve generated some embeddings, you can use the built-in utility functions to compare how similar one sentence is to another, ***including synonyms!*** 🤯 One way to do this is with the [“Cosine Similarity”](https://www.machinelearningplus.com/nlp/cosine-similarity/) function. With ST, you can skip all the pesky math and call the *very* handy `util.cos_sim` function to get a score from -1 to 1 that signifies how “similar” the embedded sentences are in the vector space they share – the bigger the score is, the more similar the sentences are! 
```

<figure class="image table text-center m-0 w-full">
  <img style="border:none;" alt="A flowchart showing sentences being embedded with Sentence Transformers, and then compared with Cosine Similarity" src="assets/83_st_first_project/sentence-transformers-explained.svg" />
  <figcaption>After embedding sentences, we can compare them with Cosine Similarity.</figcaption>
</figure>

> Being able to compare sentences by similarity means that if we have a collection of sentences or paragraphs, we can quickly find the ones that match a particular search query with a process called *[semantic search](https://www.sbert.net/examples/applications/semantic-search/README.html)*.

Reviewer: For more concrete examples, I add previous blog posts. However, feel free to disregard them.

```suggestion
Comparing sentences by similarity means that if we have a collection of sentences or paragraphs, we can quickly find the ones that match a particular search query with a process called *[semantic search](https://www.sbert.net/examples/applications/semantic-search/README.html)*. For specific applications, [this post](https://huggingface.co/spaces/sentence-transformers/Sentence_Transformers_for_semantic_search) references GitHub functions and this one [builds an FAQ engine](https://huggingface.co/blog/getting-started-with-embeddings) using Sentence Transformers.
```

## Why learn to use Sentence Transformers?

> First, it offers a low-barrier way to get hands-on experience with state-of-the-art models to generate [embeddings](https://daleonai.com/embeddings-explained). I found that creating my own sentence embeddings was a powerful learning tool which helped strengthen my understanding of how modern models work with text, and it also got the creative juices flowing for ideation! Within a few minutes of loading up the `msmarco-MiniLM-L-6-v3` model in a Jupyter notebook I’d come up with a bunch of fun project ideas just from embedding some sentences and running some of ST’s utility functions on them.

Reviewer: ```suggestion
First, it offers a low-barrier way to get hands-on experience with state-of-the-art models to generate [embeddings](https://daleonai.com/embeddings-explained). I found that creating my Sentence Embeddings model was a powerful learning tool that helped strengthen my understanding of how modern models work with text, and it also got the creative juices flowing for ideation! Within a few minutes of loading up the `msmarco-MiniLM-L-6-v3` model in a Jupyter notebook, I’d come up with a bunch of fun project ideas just from embedding some sentences and running some of ST’s utility functions on them.
```

> Sentence Transformers is also an accessible entry-point to many important ML concepts that you can branch off into. For example, you can use it to learn about [clustering](https://www.sbert.net/examples/applications/clustering/README.html), [model distillation](https://www.sbert.net/examples/training/distillation/README.html), and even launch into text-to-image work with [CLIP](https://www.sbert.net/examples/applications/image-search/README.html). In fact, Sentence Transformers is so versatile that it’s skyrocketed to almost 8,000 stars on GitHub, with [more than 3,000 projects and packages depending on it](https://github.com/UKPLab/sentence-transformers/network/dependents?dependent_type=REPOSITORY&package_id=UGFja2FnZS00ODgyNDAwNzQ%3D). On top of the official docs, there’s an abundance of community-created content (look for some links at the end of this post 👀), and the library’s ubiquity has made it [popular in research](https://twitter.com/NimaBoscarino/status/1535331680805801984?s=20&t=gd0BycVE-H4_10G9w30DcQ).

Reviewer: ```suggestion
Second, Sentence Transformers is an accessible entry-point to many important ML concepts that you can branch off into. For example, you can use it to learn about [clustering](https://www.sbert.net/examples/applications/clustering/README.html), [model distillation](https://www.sbert.net/examples/training/distillation/README.html), and even launch into text-to-image work with [CLIP](https://www.sbert.net/examples/applications/image-search/README.html). In fact, Sentence Transformers is so versatile that it’s skyrocketed to almost 8,000 stars on GitHub, with [more than 3,000 projects and packages depending on it](https://github.com/UKPLab/sentence-transformers/network/dependents?dependent_type=REPOSITORY&package_id=UGFja2FnZS00ODgyNDAwNzQ%3D). On top of the official docs, there’s an abundance of community-created content (look for some links at the end of this post 👀), and the library’s ubiquity has made it [popular in research](https://twitter.com/NimaBoscarino/status/1535331680805801984?s=20&t=gd0BycVE-H4_10G9w30DcQ).
```

> 

Reviewer: Adding what I think is a key reason for learning Sentence Transformers, feel free to disregard it if necessary 😊.

```suggestion

Third, embeddings, such as the ones generated by Sentence Transformers, are key for several industrial applications. "Google Search uses embeddings to [match text to text and text to images](https://cloud.google.com/blog/topics/developers-practitioners/meet-ais-multitool-vector-embeddings); Snapchat uses them to "[serve the right ad to the right user at the right time](https://eng.snap.com/machine-learning-snap-ad-ranking)"; and Meta (Facebook) uses them for [their social search](https://research.facebook.com/publications/embedding-based-retrieval-in-facebook-search/)" - [Getting Started with Embeddings](https://huggingface.co/blog/getting-started-with-embeddings). In other words, embeddings allow you to build everything from chatbots, recommendation systems, zero-shot classifiers, image search, and FAQ systems.

```

> On top of it all, it’s also supported with a ton of  [Hugging Face integrations](https://huggingface.co/docs/hub/sentence-transformers) 🤗

Reviewer: ```suggestion
On top of it all, it’s also supported with a ton of  [Hugging Face integrations](https://huggingface.co/docs/hub/sentence-transformers) 🤗.
```

---

Once you’ve gone through your first project, you’ll find that you’ll have even more ideas for things to work on! Have fun, and don’t forget to share your projects and everything you’ve learned with us over at [hf.co/join/discord](http://hf.co/join/discord) 🤗

Further reading:

> - [Sentence Transformers x Hugging Face](https://huggingface.co/docs/hub/sentence-transformers)

Reviewer: ```suggestion
- [Sentence Transformers organization in the Hugging Face Hub](https://huggingface.co/docs/hub/sentence-transformers)
```

- [Sentence_Transformers for Semantic Search - by Omar Espejel](https://huggingface.co/spaces/sentence-transformers/Sentence_Transformers_for_semantic_search)

> - [Pinecone.io - Sentence Embeddings](https://www.pinecone.io/learn/sentence-embeddings/#some-context)

Reviewer: Just adding the previous post.

```suggestion
- [Pinecone.io - Sentence Embeddings](https://www.pinecone.io/learn/sentence-embeddings/#some-context)
- [Getting Started with Embeddings](https://huggingface.co/blog/getting-started-with-embeddings)
```
